exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinBatch.java

```diff
@@ -465,82 +785,460 @@ public HashJoinProbe setupHashJoinProbe() throws ClassTransformationException, I
+    final StringBuilder sb = new StringBuilder();
+
+    for (int partitionIndex = 0; partitionIndex < partitions.length; partitionIndex++) {
+      final String partitionPrefix = "Partition " + partitionIndex + ": ";
+      final HashPartition hashPartition = partitions[partitionIndex];
+      sb.append(partitionPrefix).append(hashPartition.makeDebugString()).append("\n");
     }
-    htable.getStats(htStats);
-    stats.setLongStat(Metric.NUM_BUCKETS, htStats.numBuckets);
-    stats.setLongStat(Metric.NUM_ENTRIES, htStats.numEntries);
-    stats.setLongStat(Metric.NUM_RESIZING, htStats.numResizing);
-    stats.setLongStat(Metric.RESIZING_TIME_MS, htStats.resizingTime);
+
+    return sb.toString();
+  }
+
+  /**
+   * Updates the {@link HashTable} and spilling stats after the original build side is processed.
+   *
+   * Note: this does not update all the stats. The cycleNum is updated dynamically in {@link #innerNext()} and the total bytes
+   * written is updated at close time in {@link #cleanup()}.
+   */
+  private void updateStats() {
+    if ( buildSideIsEmpty ) { return; } // no stats when the right side is empty
+    if ( cycleNum > 0 ) { return; } // These stats are only for before processing spilled files
+
+    final HashTableStats htStats = new HashTableStats();
+    long numSpilled = 0;
+    HashTableStats newStats = new HashTableStats();
+    // sum the stats from all the partitions
+    for ( HashPartition partn : partitions ) {
+      if ( partn.isSpilled() ) { numSpilled++; }
+      partn.getStats(newStats);
+      htStats.addStats(newStats);
+    }
+
+    this.stats.setLongStat(Metric.NUM_BUCKETS, htStats.numBuckets);
+    this.stats.setLongStat(Metric.NUM_ENTRIES, htStats.numEntries);
+    this.stats.setLongStat(Metric.NUM_RESIZING, htStats.numResizing);
+    this.stats.setLongStat(Metric.RESIZING_TIME_MS, htStats.resizingTime);
+    this.stats.setLongStat(Metric.NUM_PARTITIONS, numPartitions);
+    this.stats.setLongStat(Metric.SPILL_CYCLE, cycleNum); // Put 0 in case no spill
+    this.stats.setLongStat(Metric.SPILLED_PARTITIONS, numSpilled);
   }
 
   @Override
   public void killIncoming(boolean sendUpstream) {
-    left.kill(sendUpstream);
-    right.kill(sendUpstream);
+    wasKilled = true;
+    probeBatch.kill(sendUpstream);
+    buildBatch.kill(sendUpstream);
   }
 
   @Override
   public void close() {
-    if (hjHelper != null) {
-      hjHelper.clear();
+    if ( cycleNum > 0 ) { // spilling happened
+      // In case closing due to cancellation, BaseRootExec.close() does not close the open
+      // SpilledRecordBatch "scanners" as it only knows about the original left/right ops.
+      killIncoming(false);
+    }
+    this.cleanup();
+    super.close();
+  }
+
+  // ==============================================================
+  //
+  //    Methods used for the probe
+  //
+  // ============================================================
+  private BatchSchema probeSchema;
+
+  enum ProbeState {
```

Comment:
I am not sure what the moving of all the probe code from HashJoinProbeTemplate to here is buying us.  I am not seeing any advantage and in fact it is odd to see member variables declared at the end of the class.  Further, I  think we would want to keep the option of having generated code for filter evaluation open for hash joins, so having the encapsulation of the HashJoinProbeTemplate is a good thing.

Reply:
Have you given this any further thought ?  Most of the complex operators in Drill are organized by way of 'Batch' and 'Template' and this structure has been well understood by most developers historically and may even be documented somewhere.   I am in favor of keeping this structure for hash join even though after your current changes the template code is not doing code gen.  I am open to getting feedback from other drill developers on what they think.